c/enc/hash_longest_match64_inc.h

```diff
@@ -177,90 +179,135 @@ static BROTLI_INLINE void FN(FindLongestMatch)(
+  while (*back_refs_position < back_refs_size &&
+         backward_references[*back_refs_position].position < cur_ix) {
+      ++(*back_refs_position);
   }
-  {
-    const uint32_t key = FN(HashBytes)(
-        &data[cur_ix_masked], self->hash_mask_, self->hash_shift_);
-    uint32_t* BROTLI_RESTRICT bucket = &buckets[key << self->block_bits_];
-    const size_t down =
-        (num[key] > self->block_size_) ?
-        (num[key] - self->block_size_) : 0u;
-    for (i = num[key]; i > down;) {
-      size_t prev_ix = bucket[--i & self->block_mask_];
-      const size_t backward = cur_ix - prev_ix;
+  /* If we have some backward reference from decoder for this position
+     check it first */
+  if (back_refs_size != 0 && *back_refs_position < back_refs_size &&
+     backward_references[*back_refs_position].position == cur_ix) {
+     FindBackwardReferenceFromDecoder(data, ring_buffer_mask,
+                                     cur_ix, cur_ix_masked, max_length,
+                                     backward_references, back_refs_position,
+                                     back_refs_size, out);
+     if (out->used_stored) {
+       return;
+     }
+  }
+  else {
+    /* Try last distance first. */
+    for (i = 0; i < (size_t)self->num_last_distances_to_check_; ++i) {
+      const size_t backward = (size_t)distance_cache[i];
+      size_t prev_ix = (size_t)(cur_ix - backward);
+      if (prev_ix >= cur_ix) {
+        continue;
+      }
       if (BROTLI_PREDICT_FALSE(backward > max_backward)) {
-        break;
+        continue;
       }
       prev_ix &= ring_buffer_mask;
+
       if (cur_ix_masked + best_len > ring_buffer_mask ||
           prev_ix + best_len > ring_buffer_mask ||
           data[cur_ix_masked + best_len] != data[prev_ix + best_len]) {
         continue;
       }
       {
-        const size_t len = FindMatchLengthWithLimit(&data[prev_ix],
-                                                    &data[cur_ix_masked],
-                                                    max_length);
-        if (len >= 4) {
-          /* Comparing for >= 3 does not change the semantics, but just saves
-             for a few unnecessary binary logarithms in backward reference
-             score, since we are not interested in such short matches. */
-          score_t score = BackwardReferenceScore(len, backward);
+        size_t len = FindMatchLengthWithLimit(&data[prev_ix],
+                                              &data[cur_ix_masked],
+                                              max_length);
+        /* If have a backward_references array cut the copy_len so
+           it won't intersect with any other stored references */
+        if (back_refs_size > 0 && *back_refs_position < back_refs_size &&
+            cur_ix + len > backward_references[*back_refs_position].position) {
+          len = backward_references[*back_refs_position].position - cur_ix;
+        }
+        if (len >= 3 || (len == 2 && i < 2)) {
```

Comment:
Why 3 and 2?

Reply:
Although it looks like my code it's not mine. I only placed it inside else branch and added 225-228 lines to cut a copy_len

Reply:
Again we limit the amount of unnecessary calculations.
`i` refers to special copy commands that are cheaper.

Reply:
If it's existing code, that's out-of-scope for the review :)